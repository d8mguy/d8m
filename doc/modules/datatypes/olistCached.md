## OlistCached

`OlistCached` is a variant of `olist`; it makes the same assumptions about its extraction function as `olist`; see that documentation for details. What's different is an internal cache named `recent`; the full list in this cache plus the slot named `official`. Code to merge `recent` into `official`, either on demand or when `recent` gets full. There's a fairly huge advantage to this when inserting new elements, since inserting a single element into an olist requires moving on average half of the existing list (assuming elements are entered in random order). With the cache, you insert into a small olist (`recent`) and only change the large olist (`official`) when flushing. This is done in a single pass through `official` which inserts all the new elements in the correct place. 

On the other hand, lookup in general requires searching in two olists, which takes longer than searching in one. If you have a usage pattern with a lot of insertions followed by a lot of lookups, a flush operation at the beginning of the lookups phase should get lookup timing very similar to a regular olist while retaining the advantage of cached insertions. 

Removing elements from an `olistCached` takes as long or longer than removing from an `olist`, unless the removed elements are likely to be in the cache (i.e. recently inserted). Thus, the potential advantage of `olistCached` is mainly for insertion. 

The cache size is part of the type, that is, is an argument to the typegen. This would be easy to change. 

The methods of `olistCached` closely match those of `olist`, with the addition of `flush`. 

#### typegen olistCached

    [olistCached(...): t1, t2, ...]
where T:: entity, and `t1`, `t2`: `T`.

`OlistCached` is enumerable via eachstart.

    val olistCached = \typegen(T::entity, xfn:\(T)->U, thresh: integer) {
        method $in = \(elt: T) -> boolean {...}
        method keyIn = \(oelt: U) -> boolean {...}
        method get = \(oelt: U) -> nilPossible(T) {...}
        method insert = \mod(elt: T) -> nothing {...}
        method insertIfUniq = \mod(elt: T) -> nothing {...}
        method count = \() -> integer {...}
        method remove = \mod(elt: T) -> nothing {...}
        method flush = \mod() -> nothing {...}
        method minelt = \() -> nilPossible(T) {...}
        method maxelt = \() -> nilPossible(T) {...}
        method popf = \mod() -> T {...}
        method popb = \mod() -> T {...}
    }
Method `$in`: this lookup code should be almost as fast as a standard `olist` when `recent` is empty. But if lookups to recent insertions are common, it will be slower than a code that tests `recent` first.

Method `keyIn`: is like `$in` but only tests the key

Method `get`: returns the (first) elt whose `oextract` is `oelt`, or nil.

Method `insert`: inserts `elt` into `recent`, after checking whether `recent.count` `==` `thresh` and flushing it if so.

Method `insertIfUniq`: same as `insert` but checks first.

Method `count`: returns number of elements in both. 

Method `remove`: removes `elt`; no indicator if `elt` not present. 

Method `minelt`: returns minimum element or nil if `[]`.

Method `maxelt`: returns maximum element or nil if `[]`. 

Method `popf`: removes and returns minimum element.

Method `popb`: removes and returns maximum element. 




