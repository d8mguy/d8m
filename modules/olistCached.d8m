// OlistCached is a
// variant of ordered lists; has internal cache named recent. Full list in slot named official. Code to merge recent
// into official, either on demand or when recent gets full. The timing behavior is somewhat complex; it's likely a net
// loss for most uses. If the usage pattern involves lookups of recently added elements, then it's likely to win.
//
// The cache size is part of the type, that is, is an arg to the typegen. I'm not sure this is a good decision. It could
// certainly go the other way.
//
// Note: old comments below, to be extracted from then deleted.
// Two olists named official and recent, both represented internally as list(T) and stored according to the ordering.
// Insert adds to recent and merges this "cache" to official when its size exceeds the value of the thresh attribute.
// This means that $in has to test both recent and official so will generally be slower than straight olist.
// However, insert will almost always be faster and even the amortized time should be faster, since moving a large
// number of items from recent to official is done in a single pass and has better cache behavior than adding the same
// number of items at random places in the large olist. Delete will only be faster if the item is found in recent; so
// another advantageous case for this version is if recently added items are accessed and/or deleted more than older ones.
//
// The semantics of ordering is the same as for olist; see the extensive discussion there for details. The short version
// is that the typegen parameter here called `xfn` should create a total order on `T`.

export olistCached

given(U::ordered) {
	val olistCached = \typegen(T::entity, xfn:\(T)->U, thresh: integer) {
		extend tuple(private official, recent: list(T), private limit: integer) where {
			// binsearch ftn that'll be used on both official and recent: returns index of elt <= oelt or -1 if no such elt
			val indexA = \imp(lst: list(T), oelt: U) {
				if(lst == [] || xfn(lst.last) < oelt) { -1 }
				else {
					var mn = 0, mx = lst.count-1, cur = mx/2
					while(mn <= mx) {
						val ocur = xfn(lst[cur])
						if(ocur < oelt) mn = cur+1
						else if(ocur == oelt) break
						else mx = cur-1
						cur = mn + (mx - mn)/2
					}
					cur
				}
			}
			// Integrates recent into official; use cleverness to ensure elts of official are moved at most once.
			private method list2official = \mod() {
			    val olenb4 = official.count
				official.insert(official.count, recent)
			    if(olenb4 > 0 && recent.count > 0) {
                    var inxo = olenb4 - 1, inxl = recent.count - 1
                    var destinx = inxo + inxl + 1
                    var oelt = xfn(official[inxo]), relt = xfn(recent[inxl])
                    loop {
                        if(oelt < relt) {
                            official[destinx] = recent[inxl]
                            inxl -= 1
                            if(inxl < 0) break
                            relt = xfn(recent[inxl])
                        } else {
                            official[destinx] = official[inxo]
                            inxo -= 1
                            if(inxo < 0) break
                            oelt = xfn(official[inxo])
                        }
                        destinx -= 1
                    }
                    if(inxl >= 0) copyFrom(official, 0, recent.select(0, inxl+1))
                }
                recent = recent.select(0,0)
/*
                var ordCount = 0
                var curval = xfn(official[0])
                var badinx = 0
                each(i^1...official.count) { val tmp = xfn(official[i]); if(curval < tmp) ordCount += 1 else badinx = i; curval = tmp }
                unless(recent.count == 0 && ordCount == official.count - 1) {
                    println("rc=", recent.count, ";oc=", ordCount, ";bi=", badinx)
                }
*/
			}

			// This lookup code should be almost as fast as a std olist when recent is empty. But if lookups to recent
			// insertions are common, it will be slower than a code that tests recent first.
			method $in = \(elt: T) -> boolean {
				val oelt = xfn(elt)
				val inx = indexA(official, oelt)
				if(inx >= 0 && official[inx] == elt) true
				else if(recent.count > 0) {
					val rinx = indexA(recent, oelt)
					rinx >= 0 && recent[rinx] == elt
				} else false
			}

            // Is like $in but only tests the key
            method keyIn = \(oelt: U) -> boolean {
                val inx = indexA(recent, oelt)
                if(inx >= 0 && xfn(recent[inx]) == oelt) true
                else {
                    val oinx = indexA(official, oelt)
                    oinx >= 0 && xfn(official[oinx]) == oelt
                }
            }

            // Returns the (first) elt whose oextract is oelt, or nil
            method get = \(oelt: U) -> nilPossible(T) {
                val inx = indexA(recent, oelt)
                if(inx >= 0 && xfn(recent[inx]) == oelt) recent[inx]
                else {
                    val oinx = indexA(official, oelt)
                    if(oinx >= 0 && xfn(official[oinx]) == oelt) official[oinx] else nil
                }
            }

			// inserts elt into recent, after checking whether recent.count == thresh and flushing it if so
			method insert = \mod(elt: T) -> nothing {
				if(recent.count >= limit) list2official()
				val inx = indexA(recent, xfn(elt))
				if(inx < 0) recent.pushb(elt)
				else recent.insert(inx, elt)
			}

            // same as insert but checks first.
			method insertIfUniq = \mod(elt: T) -> nothing {
				unless(elt in self) insert(elt)
			}

			method lithook = \mod(elts: list(T)) {
				recent = []
				official = []
				if(thresh == 0) exit("olistCached cannot be initialized with cache size = 0")
				limit = thresh
				each(e^elts) insert(e)
			}

            // is defined to support stmapping from set to olist.
            method lithook = \mod(elts: list(T), noDups: boolean) {
                if(noDups) each(e^elts) self.insertIfUniq(e)
                self
            }

			method count = \() -> integer { official.count + recent.count }
			method remove = \mod(elt: T) -> nothing {
				val oelt = xfn(elt)
				var inx = indexA(recent, oelt)
				var done = false
				if(inx >= 0) {
				    while(inx < recent.count && xfn(recent[inx]) == oelt) {
				        if(recent[inx] == elt) { recent.removeIndex(inx); done = true; break }
				        inx += 1
				    }
				}
				unless(done) {
					inx = indexA(official, oelt)
					if(inx >= 0) {
                        while(inx < official.count && xfn(official[inx]) == oelt) {
                            if(official[inx] == elt) { official.removeIndex(inx); break }
                            inx += 1
                        }
					}
				}
			}
			// Empties recent, aka "flushes the cache". This is experimental, thought to be possibly helpful for tuning
			// when we're about to do a lot of lookups. Probably overkill in fact.
			method flush = \mod() -> nothing { list2official() }
			method eachstart = \imp() {
				val theOls = self		// have to refnc the type inside the nested type
				val estype = extend tuple(inxr, inxo: integer, onRecent: boolean) where {
					method value = \mod() {
					    onRecent = (inxo >= theOls.official.count || xfn(theOls.recent[inxr]) <= xfn(theOls.official[inxo]))
					    onRecent ? theOls.recent[inxr] : theOls.official[inxo]
					}
					method eachdone = \() { inxr >= theOls.recent.count && inxo >= theOls.official.count }
					method eachstep = \mod() {
						if(onRecent) inxr += 1 else inxo += 1
					}
				}
				// the eachstart prcl requires eachstep to return nothing, which is inconvenient in this case
				[estype: 0, 0, false]
			}
			method minelt = \() -> nilPossible(T) {
			    if(self.count == 0) nil
			    else if(recent.count == 0) official[0]
			    else if(official.count == 0) recent[0]
			    else {
			        val orcnt = xfn(recent[0]), oocnt = xfn(official[0])
			        orcnt < oocnt ? recent[0] : official[0]
			    }
			}
			method maxelt = \() -> nilPossible(T) {
			    if(self.count == 0) nil
			    else if(recent.count == 0) official[official.count - 1]
			    else if(official.count == 0) recent[recent.count - 1]
			    else {
			        val orcnt = xfn(recent[recent.count - 1]), oocnt = xfn(official[official.count - 1])
			        oocnt < orcnt ? recent[recent.count - 1] : official[official.count - 1]
			    }
			}
			method popf = \mod() -> nilPossible(T) {
			    if(self.count == 0) nil
                else if(recent.count == 0) official.popf()
                else if(official.count == 0) recent.popf()
                else {
                    val orcnt = xfn(recent[0]), oocnt = xfn(0)
                    orcnt < oocnt ? recent.popf : official.popf
                }
			}
			method popb = \mod() -> nilPossible(T) {
			    if(self.count == 0) nil
                else if(recent.count == 0) official.popb
                else if(official.count == 0) recent.popb
                else {
                    val orcnt = xfn(recent[recent.count - 1]), oocnt = xfn(official.count - 1)
                    oocnt < orcnt ? recent.popb : official.popb
                }
			}
		}
	}
}

oncondition defined(set, type) {
    given(T0::entity, OT::ordered, ox:\(T0)->OT, top: integer, aT: T0, aLT: list(T0)) {
        STMap {
            source = set(T0)
            target = olistCached(T0, ox, top)
            direct count, $in
            lithook(aLT) => lithook(aLT, true)
            add(aT) => insertIfUniq(aT)
    		remove(aT) => remove(aT)
        }
    }
}
