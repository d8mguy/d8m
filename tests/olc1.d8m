import "olistCached" as olc
import go "math/rand"
val tuptest = tuple(x, y: integer)
val olcI = olc.olistCached(tuptest, \(tp:tuptest) { tp.y }, 100)
val tryolc = \imp() {
    val randx = rand.New(rand.NewSource(1))
    var olc1 = [olcI: ]
    each(i^1..400) olc1.insert([tuptest: randx.Intn(600), randx.Intn(3000)])
    olc1.{this}
}
tryolc()
