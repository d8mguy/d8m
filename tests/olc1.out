[[340; 0], [1; 6], [359; 7], [546; 9], [247; 10], [352; 11], [119; 14], [390; 15], [574; 39], [113; 45], [522; 45], [47; 59], [300; 60], [589; 72], [466; 75], [120; 81], [485; 88], [162; 89], [201; 90], [294; 98], [451; 121], [337; 133], [499; 138], [331; 146], [433; 147], [317; 151], [96; 152], [405; 156], [423; 162], [599; 162], [75; 181], [215; 185], [326; 196], [514; 207], [313; 214], [529; 231], [367; 231], [174; 237], [362; 243], [520; 260], [466; 270], [509; 277], [262; 283], [320; 289], [256; 300], [122; 301], [67; 301], [481; 318], [86; 320], [587; 326], [345; 339], [292; 342], [111; 343], [148; 350], [304; 356], [429; 356], [124; 357], [244; 364], [393; 380], [415; 385], [221; 390], [17; 393], [323; 394], [541; 408], [590; 417], [4; 421], [51; 434], [395; 450], [63; 454], [561; 460], [577; 463], [338; 472], [6; 474], [506; 483], [482; 495], [527; 496], [251; 510], [494; 511], [248; 513], [430; 513], [92; 524], [114; 527], [412; 532], [459; 546], [498; 553], [65; 558], [133; 561], [504; 562], [221; 574], [431; 577], [598; 585], [537; 631], [586; 632], [239; 637], [40; 652], [328; 653], [240; 662], [390; 666], [95; 666], [65; 682], [4; 685], [53; 701], [85; 710], [216; 719], [357; 721], [323; 723], [144; 723], [8; 727], [514; 730], [169; 740], [383; 746], [501; 750], [118; 756], [556; 765], [552; 775], [412; 779], [320; 783], [437; 800], [326; 802], [303; 804], [298; 817], [128; 818], [416; 839], [93; 840], [552; 843], [521; 855], [559; 861], [493; 870], [503; 888], [471; 894], [384; 903], [395; 914], [141; 915], [246; 921], [210; 921], [583; 928], [175; 943], [19; 981], [270; 984], [451; 985], [8; 1008], [287; 1010], [173; 1017], [285; 1026], [322; 1029], [584; 1044], [36; 1051], [401; 1058], [461; 1060], [244; 1063], [251; 1076], [235; 1079], [220; 1079], [472; 1088], [13; 1090], [429; 1092], [474; 1099], [437; 1106], [409; 1128], [63; 1135], [84; 1150], [253; 1158], [104; 1162], [78; 1175], [370; 1182], [305; 1183], [395; 1197], [8; 1198], [19; 1200], [573; 1201], [161; 1202], [380; 1205], [288; 1223], [358; 1232], [128; 1258], [328; 1274], [172; 1275], [355; 1277], [10; 1285], [324; 1293], [384; 1297], [351; 1312], [478; 1324], [359; 1325], [201; 1328], [131; 1332], [470; 1336], [278; 1336], [436; 1338], [521; 1350], [40; 1351], [225; 1351], [559; 1353], [103; 1355], [100; 1359], [595; 1362], [320; 1374], [544; 1395], [393; 1398], [477; 1402], [459; 1414], [482; 1414], [202; 1418], [552; 1420], [335; 1440], [35; 1443], [411; 1445], [516; 1452], [354; 1464], [86; 1467], [454; 1483], [236; 1487], [441; 1490], [194; 1510], [413; 1516], [27; 1534], [106; 1537], [315; 1540], [24; 1547], [60; 1551], [394; 1563], [188; 1571], [233; 1574], [164; 1588], [514; 1589], [79; 1592], [539; 1593], [205; 1598], [88; 1611], [330; 1619], [48; 1630], [233; 1643], [311; 1648], [587; 1657], [39; 1694], [156; 1695], [400; 1705], [1; 1707], [366; 1710], [369; 1712], [2; 1718], [354; 1721], [275; 1740], [506; 1740], [580; 1741], [283; 1755], [48; 1756], [381; 1757], [315; 1757], [105; 1762], [53; 1795], [294; 1806], [89; 1814], [580; 1834], [101; 1847], [345; 1871], [310; 1872], [175; 1874], [456; 1878], [534; 1879], [281; 1887], [487; 1888], [191; 1898], [29; 1920], [423; 1953], [36; 1955], [70; 1996], [491; 2002], [399; 2024], [325; 2026], [10; 2030], [171; 2039], [420; 2044], [573; 2051], [41; 2059], [544; 2078], [381; 2079], [137; 2086], [293; 2086], [95; 2092], [314; 2093], [447; 2094], [546; 2107], [414; 2118], [581; 2131], [467; 2137], [392; 2138], [201; 2140], [163; 2143], [178; 2154], [413; 2162], [365; 2162], [53; 2166], [284; 2173], [252; 2175], [106; 2179], [157; 2184], [225; 2190], [189; 2199], [328; 2201], [429; 2203], [342; 2208], [521; 2213], [509; 2215], [163; 2217], [253; 2218], [107; 2222], [222; 2223], [445; 2227], [253; 2240], [177; 2242], [202; 2258], [177; 2262], [77; 2267], [405; 2276], [178; 2281], [574; 2284], [247; 2292], [376; 2300], [182; 2316], [364; 2339], [353; 2360], [222; 2371], [163; 2376], [518; 2392], [208; 2407], [217; 2407], [277; 2412], [457; 2415], [196; 2420], [307; 2428], [487; 2430], [1; 2438], [223; 2442], [484; 2442], [200; 2456], [587; 2462], [309; 2466], [95; 2466], [68; 2476], [249; 2488], [540; 2503], [587; 2532], [488; 2538], [25; 2540], [64; 2560], [2; 2564], [131; 2574], [391; 2582], [573; 2596], [516; 2600], [599; 2615], [156; 2617], [408; 2631], [390; 2632], [271; 2641], [541; 2642], [437; 2647], [6; 2648], [123; 2650], [157; 2652], [114; 2658], [151; 2666], [219; 2666], [45; 2671], [130; 2686], [312; 2695], [589; 2698], [513; 2709], [201; 2722], [123; 2724], [133; 2740], [387; 2749], [399; 2760], [425; 2768], [53; 2774], [140; 2786], [277; 2801], [559; 2807], [19; 2807], [335; 2815], [466; 2828], [587; 2831], [260; 2834], [345; 2835], [560; 2836], [281; 2848], [340; 2860], [459; 2867], [374; 2869], [579; 2875], [79; 2887], [294; 2922], [516; 2926], [352; 2928], [447; 2947], [543; 2968], [273; 2972], [494; 2985], [237; 2998]]
